src/backend/distributed/executor/executor.c

```diff
@@ -1403,6 +1512,84 @@ ConnectionStateMachine(WorkerSession *session)
 }
 
 
+/*
+ * Activate2PCIfModifyingTransactionExpandsToNewNode sets the coordinated
+ * transaction to use 2PC under the following circumstances:
+ *     - We're already in a transaction block
+ *     - At least one of the previous commands in the transaction block
+ *       made a modification, which have not set 2PC itself because it
+ *       was a single shard command
+ *     - The input "session" is used for a distributed execution which
+ *       modifies the database. However, the session (and hence the
+ *       connection) is established to a different worker than the ones
+ *       that is used previously in the transaction.
+ *
+ *  To give an example,
+ *      BEGIN;
+ *          -- assume that the following INSERT goes to worker-A
+ *          -- also note that this single command does not activate
+ *          -- 2PC itself since it is a single shard mofication
+ *          INSERT INTO distributed_table (dist_key) VALUES (1);
+ *
+ *          -- do one more single shard UPDATE hitting the same
+ *          shard (or worker node in general)
+ *          -- this wouldn't activate 2PC, since we're operating on the
+ *          -- same worker node that we've modified earlier
+ *          -- so the executor would use the same connection
+ *			UPDATE distributed_table SET value = 10 WHERE dist_key = 1;
+ *
+ *          -- now, do one more INSERT, which goes to worker-B
+ *          -- At this point, this function would activate 2PC
+ *          -- since we're now expanding to a new node
+ *          -- for example, if this command were a SELECT, we wouldn't
+ *          -- activate 2PC since we're only interested in modifications/DDLs
+ *          INSERT INTO distributed_table (dist_key) VALUES (2);
+ */
+static void
+Activate2PCIfModifyingTransactionExpandsToNewNode(WorkerSession *session)
+{
+	DistributedExecution *execution = NULL;
+
+	if (MultiShardCommitProtocol != COMMIT_PROTOCOL_2PC ||
+		DoesCoordinatedTransactionUse2PC())
+	{
+		/* we don't need 2PC or we're already using 2PC, so no need to continue */
+		return;
+	}
+
+	execution = session->workerPool->distributedExecution;
+	if (TransactionModifiedDistributedTable(execution) &&
+		DistributedPlanModifiesDatabase(execution->plan) &&
+		!ConnectionModifiedPlacement(session->connection))
```

Comment:
Calling `ConnectionModifiedPlacement` gets inefficient for long transaction blocks that do mostly SELECTs. That might be acceptable, but still wondering whether there is a better place to call this function.

Reply:
But this is called after `TransactionModifiedDistributedTable(execution) && DistributedPlanModifiesDatabase(execution->plan)`, so shouldn't affect select only workloads?

Reply:
Oh, sorry you're mentioning `mostly single-tenant SELECT`, let me think a bit more

Reply:
I think we should look for a better place to call it from, so that it doesn't get called quite so often. e.g. in StartPlacementExecutionOnSession?

Reply:
`StartPlacementExecutionOnSession ` is a little bit trick, because when `SET citus.force_max_query_parallelization TO ON;`, we're calling `StartPlacementListConnection`, which sets the placement accesses even before the connection is established.

Can you check where I added the call now?